src/uploaders.rs

```diff
@@ -0,0 +1,147 @@
+use conduit::Request;
+use krate::Crate;
+use util::{CargoResult, internal, ChainError};
+use util::{LimitErrorReader, HashingReader, read_le_u32};
+use s3;
+use semver;
+use app::{App, RequestApp};
+use std::sync::Arc;
+use std::fs::{self, File};
+use std::env;
+use std::io;
+
+#[derive(Clone)]
+pub enum Uploader {
+    /// For production usage, uploads and redirects to s3.
+    /// For test usage with a proxy.
+    S3 { bucket: s3::Bucket, proxy: Option<String> },
+
+    /// For development usage only: "uploads" crate files to `dist` and serves them
+    /// from there as well to enable local publishing and download
+    Local,
+
+    /// For one-off scripts where creating a Config is needed, but uploading is not.
+    NoOp,
+}
+
+impl Uploader {
+    pub fn proxy(&self) -> Option<&str> {
+        match *self {
+            Uploader::S3 { ref proxy, .. } => proxy.as_ref().map(String::as_str),
+            Uploader::Local | Uploader::NoOp => None,
+        }
+    }
+
+    pub fn crate_location(&self, crate_name: &str, version: &str) -> String {
+        match *self {
+            Uploader::S3 { ref bucket, .. } => {
+                format!("https://{}/{}",
+                        bucket.host(),
+                        Uploader::crate_path(crate_name, version))
+            },
+            Uploader::Local => {
+                format!("/local_uploads/{}",
+                        Uploader::crate_path(crate_name, version))
+            }
+            Uploader::NoOp => panic!("no-op uploader does not have crate files"),
```

Comment:
Do you think it's worth returning an `Option` here so that callers are a bit more explicit when they require this?